### PR TITLE
 [FLINK-28848][table-planner] Introduces LOOKUP join hint to support delayed retry for lookup join (table alias unsupported in hint)

### PR DESCRIPTION
**Pull request by lincoln-lil**
## What is the purpose of the change
Introduces LOOKUP join hint to support delayed retry for lookup join (table alias unsupported in hint)
This is the main part of FLINK-28779 to implement FLIP-234: Support Retryable Lookup Join To Solve Delayed Updates Issue In External Systems.
The join hint based on [FLINK-28682][table-planner] support join hint in batch rules #20359, and adds a new hint 'LOOKUP' described in FLIP-234.

## Brief change log
* Add new hint LOOKUP
* Add related option class and validation
* Add related tests for new classes
* Adapt to async operator with retry and also adds a new RetryableLookupFunctionDelegator to wrap user's sync lookup func to support retry

## Verifying this change
* LookupJoinHintSpecSerdeTest
* LookupJoinHintSpecTest
* LookupJoinJsonPlanTest
* LookupJoinTest
* RetryableLookupFunctionDelegatorTest
* more ITCases will be updated in [Add it cases with lookup hint using new lookup function api for lookup join](https://issues.apache.org/jira/browse/FLINK-28849)

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with @Public(Evolving): (no)
  - The serializers: (no )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation
  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by lincoln-lil** [comment]
rebased master and updated the pr

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java [hunk 1]
```diff
@@ -144,75 +146,92 @@ private LookupJoinUtil() {
...
+            }
+            if (provider instanceof AsyncLookupFunctionProvider) {
                 if (provider instanceof PartialCachingAsyncLookupProvider) {
                     PartialCachingAsyncLookupProvider partialCachingLookupProvider =
                             (PartialCachingAsyncLookupProvider) provider;
-                    return new CachingAsyncLookupFunction(
-                            partialCachingLookupProvider.getCache(),
-                            partialCachingLookupProvider.createAsyncLookupFunction());
+                    asyncLookupFunctions =
+                            new CachingAsyncLookupFunction(
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java:219]: this function should be refactored to create nested delegators of user lookup func

[CachingLookupFunction]
                   |
[RetryableLookupFunctionDelegator]
                   |
         userLookupFunc

I'll update this

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @lincoln-lil , I left some comments, and could you add some hint propagation tests ?

**Review by lincoln-lil — COMMENTED** [review]
@godfreyhe thanks for reviewing this! I'll update the pr according to your comments

**Review by godfreyhe — APPROVED** [review]
LGTM

**Comment by lincoln-lil** [comment]
@flinkbot run azure

**Comment by lincoln-lil** [comment]
@flinkbot run azure

**Comment by lincoln-lil** [comment]
there're several merge conflicts for the latest master, so need to verify the merged pr completely, the local verify are running

**Comment by lincoln-lil** [comment]
local verify ok for table modules(the ci pipeline was too slow)
image

**Pull request closed without merging**